pkg/tasker/tasker.go

```diff
@@ -0,0 +1,197 @@
+	sink := make(chan *tellor.ITellorNewChallenge)
+	var tellorFilterer *tellor.ITellorFilterer
+	tellorFilterer, err := tellor.NewITellorFilterer(mt.contractInstance.Address, mt.client)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting ITellorFilterer instance")
+	}
+	sub, err := tellorFilterer.WatchNewChallenge(&bind.WatchOpts{}, sink, nil)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting NewChallenge channel")
+	}
+	return sink, sub, nil
+}
+
+func (mt *Tasker) sendWork(challenge *tellor.ITellorNewChallenge, delay time.Duration) {
+	if challenge.CurrentRequestId[0].Int64() > int64(100) || challenge.CurrentRequestId[0].Int64() == 0 {
+		level.Warn(mt.logger).Log("msg", "new current variables request ID not correct - contract about to be upgraded")
+		return
+	}
+	newChallenge := &mining.MiningChallenge{
+		Challenge:  challenge.CurrentChallenge[:],
+		Difficulty: challenge.Difficulty,
+		RequestIDs: challenge.CurrentRequestId,
+	}
+	for _, acc := range mt.accounts {
+		level.Info(mt.logger).Log("msg", "new challenge",
+			"addr", acc.Address.String(),
+			"challenge", fmt.Sprintf("%x", newChallenge.Challenge),
+			"difficulty", newChallenge.Difficulty,
+			"requestIDs", fmt.Sprintf("%+v", newChallenge.RequestIDs),
+		)
+		mt.workSinks[acc.Address.String()] <- &mining.Work{Challenge: newChallenge, PublicAddr: acc.Address.String(), Start: uint64(rand.Int63()), N: math.MaxInt64}
+		select {
+		case <-mt.ctx.Done():
+			return
+		case <-time.After(delay):
+			continue
+		}
+	}
+
+}
+
+func (mt *Tasker) Start() error {
+	var err error
+	level.Info(mt.logger).Log("msg", "tasker has been started")
+
+	// Getting current challenge from the contract.
+	newVariables, err := mt.contractInstance.GetNewCurrentVariables(nil)
+	if err != nil {
+		level.Warn(mt.logger).Log("msg", "new current variables retrieval - contract might not be upgraded", "err", err)
+		return err
+	}
+
+	currentChallenge := &tellor.ITellorNewChallenge{
+		CurrentChallenge: newVariables.Challenge,
+		Difficulty:       newVariables.Difficutly,
+		CurrentRequestId: newVariables.RequestIds,
+		TotalTips:        newVariables.Tip,
+	}
+	if err != nil {
+		return errors.Wrap(err, "getting the current challenge")
+	}
+
+	level.Info(mt.logger).Log("msg", "sending the initial challenge to the miner")
+	mt.sendWork(currentChallenge, 5*time.Second)
+
+	// Subscribe and wait until the context cancellation event.
+	var sink chan *tellor.ITellorNewChallenge
+	var sub event.Subscription
+
+	// Initial subscription.
+	for {
+		sink, sub, err = mt.getNewChallengeChannel()
+		if err != nil {
+			level.Error(mt.logger).Log("msg", "initial subscribing to NewChallenge events failed")
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
+
+	for {
+		select {
+		case <-mt.ctx.Done():
+			level.Info(mt.logger).Log("msg", "unsubscribing from NewChallenge events")
+			if sub != nil {
+				sub.Unsubscribe()
+			}
```

Comment:
This is already done in the watch function in the binding file no?

```suggestion
```

Reply:
You're right, thanks.